### PR TITLE
Clean up and refactor Muon fitting code

### LINKED ISSUE #28531: Muon analysis fitting refactor
The fitting and sequential tabs have got very complicated. This issue is to clean up the code and make it easier to maintain.

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR removes some duplicated code in the fitting tab, which emerged when the sequential fitting tab was created. It also corrects a few behaviours which improves usability:

1. Plot guess will now update when a parameter is changed
2. When a function parameter is changed in the fitting tab the sequential fitting tab will update accordingly
3. The fitting options (at the bottom of the fitting tab) will update the model when they are changed. Previously the options in the model would only update when the function structure was changed.

In addition, some new unit tests have been added for the fitting tab model and presenter.

**To test:**
1. Load in some data (MUSR62260)
2. Go to the fitting tab
3. Add a fit function (GausOsc)
4. Check the following (amongst a number of other things):
     1. Plot guess - the guess should update if a parameter is changed)
     2. Do a single fit 
     3. Go the sequential fitting tab, the parameter for one of the fits should have updated
     3. Change start and end times and redo the fit, the range should have changed.
     4. Change to simultaneous fit and do the same steps (ii-iv)
5. In general, play around with it a bit to see if anything breaks. 

Fixes #28531 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — CHANGES_REQUESTED** [review]
Just the one thing I found in the code. Noticed a couple of other things when testing:
- Ticking the tiled plot button on the plot keeps 3 empty plots on the window. 
- The undo fits button does not always remove the fit lines from the plots when a simultaneous or sequential fit is done. 
- If tiled plot is clicked, the plot guess always appears on the first plot, not the plot holding the data. 

Everything else seems to be working as described, nice work.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py [hunk 1]
```diff
@@ -81,25 +80,87 @@ def get_function_name(self, function):
...
+            guess_ws_name = FREQUENCY_DOMAIN_ANALYSIS_GUESS_WS
+        else:
+            guess_ws_name = '__unknown_interface_fitting_guess'
+        try:
+            EvaluateFunction(InputWorkspace=data_ws_name,
+                             Function=fit_function,
+                             StartX=self.fitting_options["startX"],
+                             EndX=self.fitting_options["endX"],
+                             OutputWorkspace=guess_ws_name)
+        except RuntimeError:
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py]: Ah yes, I should have spotted that!

**Comment by StephenSmith25** [comment]
> 
> 
> Just the one thing I found in the code. Noticed a couple of other things when testing:
> 
>     * Ticking the tiled plot button on the plot keeps 3 empty plots on the window.
> 
>     * The undo fits button does not always remove the fit lines from the plots when a simultaneous or sequential fit is done.
> 
>     * If tiled plot is clicked, the plot guess always appears on the first plot, not the plot holding the data.
> 
> 
> Everything else seems to be working as described, nice work.

- So the tiled plotting behaviour only really works if you are doing a simultaneous fit and there are multiple groups/runs to plot. Annoyingly it’s a choice between an empty set of axes and a blank space where the axis should be. Probably need to look into this further in another issue, possibly to prevent tiled plotting when you’re only looking at a single fit. 
- Hmm, I assume this was after doing a fit on the sequential tab? In this case the undo fits button doesn't really work when switching between the tabs. May need to look into this in another issue to create a unified “undo fit” button which works on both tabs.
- I've changed the way the plot guess is added to the workspace, so it should now appear on the correct axis.

Linking an issue to cover these two issues #28672, which I’ll have to look at in another PR.

**Review by cailafinn — DISMISSED** [review]
Code looks good and functionality still works well. Other concerns addressed in linked issue. Nice work 👍

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py [hunk 2]
```diff
@@ -81,25 +80,74 @@ def get_function_name(self, function):
...
+        if self.fitting_options["fit_type"] == "Single":
+            if self.fitting_options["tf_asymmetry_mode"]:
                 params = self.get_parameters_for_single_tf_fit(workspace[0])
                 function_object, output_status, output_chi_squared = self.do_single_tf_fit(params)
             else:
                 params = self.get_parameters_for_single_fit(workspace[0])
                 function_object, output_status, output_chi_squared = self.do_single_fit(params)
         else:  # single simultaneous fit
-            if self.tf_asymmetry_mode:
+            if self.fitting_options["tf_asymmetry_mode"]:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py:143]: I think we can make more these text strings `suffix`

**Comment by AnthonyLim23** [comment]
Just had a go and I can cause a crash. 

1. Load MUSR 62260
2. Go to fitting and add a GausOsc
3. Make it simultaneous and share frequency and sigma
4. Set frequency to 1.3
5. Do fit
6. Turn on TF asymmetry mode
7. Do fit
8. Turn off TF asymmetry mode - you will get the below errors
9. Add a function - crash

```
Traceback (most recent call last):
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\fitting_tab_widget\fitting_tab_presenter.py", line 316, in handle_tf_asymmetry_mode_changed
    self.handle_display_workspace_changed()
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\fitting_tab_widget\fitting_tab_presenter.py", line 118, in handle_display_workspace_changed
    self.handle_plot_guess_changed()  # update the guess (use the selected workspace as data for the guess)
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\fitting_tab_widget\fitting_tab_presenter.py", line 162, in handle_plot_guess_changed
    self.model.change_plot_guess(self.view.plot_guess, workspaces, index)
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\fitting_tab_widget\fitting_tab_model.py", line 85, in change_plot_guess
    guess_ws_name = self.evaluate_plot_guess(workspace_names, plot_guess, index)
  File "C:\Users\BTR75544\work\dev\mantid\scripts\Muon\GUI\Common\fitting_tab_widget\fitting_tab_model.py", line 97, in evaluate_plot_guess
    guess_ws_name = MUON_ANALYSIS_GUESS_WS + data_ws_name
TypeError: can only concatenate str (not "NoneType") to str
```

image

**Review by martyngigg — DISMISSED** [review]
Just marking this request changes so that it doesn't get merged by accident.

**Review by AnthonyLim23 — APPROVED** [review]

**Pull request merged**